### PR TITLE
feat: introduce no_std support for ff, curve, and crypto

### PR DESCRIPTION
**Pull request by tdelabro**

### CODE REVIEW

File: starknet-crypto/src/lib.rs
Comment (xJonathanLEI) [inline starknet-crypto/src/lib.rs:13]: Do we need this to be `pub`? I think we shouldn't re-export it?

File: starknet-crypto/src/with_std.rs
Comment (xJonathanLEI) [inline starknet-crypto/src/with_std.rs:2]: The global `core` crate is available even when `no_std` is off (aka. normal mode). So we should not need this conditional import? Should be able to just change all `std::fmt` etc. to `core::fmt` unconditionally for example?

**Comment by xJonathanLEI** [comment]
Found this clippy lint that might be useful:

https://github.com/rust-lang/rust-clippy/issues/7724

File: starknet-crypto/src/with_std.rs
Comment (xJonathanLEI) [reply starknet-crypto/src/with_std.rs:2]: I think the best approach here is to apply the clippy lint(s) project-wide in `master` first, and rebase this PR? That way we don't have to manually inspect every conditional import to see if they're not needed.

Comment (xJonathanLEI) [reply starknet-crypto/src/with_std.rs:2]: Played around the idea a bit and got bitten by `thiserror` lol - enabling the lints results in every `thiserror::Error` derive getting error. So I would have to deal with `thiserror` first, which is exactly what this PR already does.

So I'll simply add the lints on top of this PR to `starknet-core` and its dependencies to make sure we're not doing unnecessary conditional imports. Then I will apply the lints to other crates in a separate PR.

(btw I'm kinda annoyed that there's no way to enable/disable lints globally in a workspace/project, not until this merges: https://github.com/rust-lang/rfcs/pull/3389)

**Comment by xJonathanLEI** [comment]
Note that we should also add CI steps for building without the `std` feature to make sure we're not accidentally breaking `no_std` support.

**Comment by xJonathanLEI** [comment]
Hmmm I just tried running this:

```console
$ cargo build --package starknet-crypto --no-default-features
```

to make sure it _really_ works for `no_std`, but I'm getting errors. Any ideas?

**Comment by xJonathanLEI** [comment]
Is it possible that we make the crate work with everything turned off (not even `alloc`), but with reduced functionality? Anything that requires `alloc` gets gated by `any(feature = "std", feature = "alloc")`. Then at least the crate would work with a bare `default-features = false`.

**Comment by tdelabro** [comment]
> Is it possible that we make the crate work with everything turned off (not even alloc), but with reduced functionality?

Yeah totally. `alloc` is only used by starknet-ff here:
```rust
impl Serialize for FieldElement {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&ToString::to_string(&self))
    }
}

impl<'de> Deserialize<'de> for FieldElement {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let value = String::deserialize(deserializer)?;
        Self::from_str(&value).map_err(serde::de::Error::custom)
    }
}
```

So what we can do is:
```toml
[dependencies]
serde = { version = "1.0.152", default-features = false, optional = true }

[features]
serde = ["alloc", "dep:serde"]
alloc = [""serde?/alloc"]
```

File: starknet-crypto/src/with_std.rs
Comment (tdelabro) [reply starknet-crypto/src/with_std.rs:2]: > Should be able to just change all std::fmt etc. to core::fmt unconditionally for example?

Yes we can. But then we have to do conditional import for symbols in `alloc`:
```rust
#[cfg(feature = "std")]
use std::rc::Rc;
#[cfg(all(not(feature = "std")), feature = "alloc")]
use alloc::rc::Rc;
```

I tried both and I really find it more comfortable to have all those stuffs managed once and for all in one place, by the library maintainer, and then anybody contributing just has to import what it needs from one single place.

It may seem a bit artificial for small repositories that don't do extensive usage of the `std`/`core`/`alloc` symbols, but in bigger repositories, it is such a quality-of-life improvement.

Nevertheless, I'm open to removing it and keeping it manual if it really bother you

Comment (tdelabro) [reply starknet-crypto/src/with_std.rs:2]: About `thiserror`. It does not support `no_std`. I used `this-error-no-std` instead, which is kind of a wrapper crate I believe. With a lot of complex macro in it, so it is possible that it does strange stuffs to the linter
`snafu` is an alternative crate that supports `no_std` natively, but it is not iso-feature with `thiserror`. It's another paradigm.

File: starknet-crypto/src/lib.rs
Comment (tdelabro) [reply starknet-crypto/src/lib.rs:13]: yep you are right

**Comment by tdelabro** [comment]
@xJonathanLEI I pushed a new version. Where the tree crates are full `#![no_std]` and don't relly at all on `std`. 
They occasionally need `alloc` to perform some specific actions that are behind feature flags, such as `serde` or `display-field-element`.

I also replaced `thiserror` by `snafu` in order to have non-buggy `no_std`.

I think you will like this version better

File: starknet-crypto/Cargo.toml
Comment (xJonathanLEI) [inline starknet-crypto/Cargo.toml:27]: Is there a specific reason `hashbrown` gets added here?

File: starknet-ff/Cargo.toml
Comment (xJonathanLEI) [inline starknet-ff/Cargo.toml:23]: As mentioned in the Telegram chat, it looks like `bigdecimal-no-std` isn't _really_ `no_std` compatible as its name suggests - I still have to enable `std` for it to compile for `thumbv6m-none-eabi`.

Maybe we simply revert back to using the original `bigdecimal` crate and make it depend on `std` instead?

File: starknet-ff/Cargo.toml
Comment (xJonathanLEI) [inline starknet-ff/Cargo.toml:28]: I think it makes more sense for `serde` to also be enabled by default.

**Review by xJonathanLEI — COMMENTED** [review]
Thanks for updating this! Looks good to me apart from a few minor issues

File: starknet-crypto/Cargo.toml
Comment (tdelabro) [reply starknet-crypto/Cargo.toml:27]: nope. It's an error

File: starknet-ff/Cargo.toml
Comment (tdelabro) [reply starknet-ff/Cargo.toml:28]: ok

File: starknet-ff/Cargo.toml
Comment (tdelabro) [reply starknet-ff/Cargo.toml]: > Maybe we simply revert back to using the original bigdecimal crate and make it depend on std instead?

Done. Good enough for now.
I will try to create a `no_std` branch for `bigdecimal` later, and maybe merge it in the original repo. But it is not really active...

File: starknet-ff/src/lib.rs
Comment (tdelabro) [reply starknet-ff/src/lib.rs:8]: totally

**Review by xJonathanLEI — APPROVED** [review]
LGTM. Thanks a lot for working on this!

**Pull request merged**
